File: packages/acp-sdk-ts/src/registry.ts
/** Component registry: id -> ComponentDef, with descriptor projection. */
import type { ComponentDef } from "./component.js";
import { toDescriptor } from "./component.js";
import type { ComponentDescriptor } from "./types.js";
import { isValidComponentId } from "./codec.js";

/** Variance-loose form used by the registry (handlers keep their precise types at definition sites). */
export type AnyComponentDef = ComponentDef<any, any>;

export class Registry {
  #components = new Map<string, AnyComponentDef>();

  register(def: AnyComponentDef): void {
    if (!isValidComponentId(def.id)) {
      throw new Error(`invalid component id: ${def.id} (spec §7.1)`);
    }
    if (typeof def.handle !== "function") {
      throw new Error(`component ${def.id}: handle must be a function`);
    }
    if (this.#components.has(def.id)) {
      throw new Error(`component already registered: ${def.id}`);
    }
    this.#components.set(def.id, def);
  }

  get(id: string): ComponentDef | undefined {
    return this.#components.get(id);
  }

  has(id: string): boolean {
    return this.#components.has(id);
  }

  list(): AnyComponentDef[] {
    return [...this.#components.values()];
  }

  descriptors(): ComponentDescriptor[] {
    return this.list().map(toDescriptor);
  }

  get size(): number {
    return this.#components.size;
  }
}

File: sdk/python/acp/registry.py
"""Component registry: id -> ComponentDef, with descriptor projection."""

from typing import Dict, List, Optional

from .codec import is_valid_component_id
from .component import ComponentDef, to_descriptor


class Registry:
    """Registered component definitions keyed by component id (spec §7.1)."""

    def __init__(self) -> None:
        self._components: Dict[str, ComponentDef] = {}

    def register(self, def_: ComponentDef) -> ComponentDef:
        if not is_valid_component_id(def_.id):
            raise ValueError("invalid component id: {} (spec §7.1)".format(def_.id))
        if not callable(def_.handle):
            raise ValueError("component {}: handle must be callable".format(def_.id))
        if def_.id in self._components:
            raise ValueError("component already registered: {}".format(def_.id))
        self._components[def_.id] = def_
        return def_

    def get(self, component_id: str) -> Optional[ComponentDef]:
        return self._components.get(component_id)

    def has(self, component_id: str) -> bool:
        return component_id in self._components

    def list(self) -> List[ComponentDef]:
        return list(self._components.values())

    def descriptors(self) -> List[dict]:
        return [to_descriptor(d) for d in self._components.values()]

    def __len__(self) -> int:
        return len(self._components)
